### PR TITLE
mixed precision training: change optimize_role ops order behind backward_role ops

### PR DESCRIPTION
**Pull request by danleifeng**
in fp16_util.py, update_role_var_grad function will change cast op role, which makes effect in parallelExecutor. But in Executor, it may cause errors without nccl synchronization. 

To solve this problem, we can move optimize_role ops behind all the backward_role ops. It can also speed up Executor training.

### CODE REVIEW

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 1]
```diff
@@ -35,6 +35,30 @@ def _rename_arg(op, old_name, new_name):
...
+    put optimize_role ops(cast) behind the backward_role ops
+    to speed up in Executor.
+
+    Args:
+        block: The global_block of main program. eg: block = main_prog.global_block()
+    """
+    OPTIMIZE = core.op_proto_and_checker_maker.OpRole.Optimize
+    for idx, op in reversed(list(enumerate(block.ops))):
+        if op.attr('op_role') & int(OPTIMIZE):
+
```
Comment (gavin1332) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: unexpected blank line

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 2]
```diff
@@ -35,6 +35,30 @@ def _rename_arg(op, old_name, new_name):
     op_desc._rename_output(old_name, new_name)
 
 
+def _change_optimize_ops_order(block):
```
Comment (gavin1332) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: _move_optimize_ops_back(block) may be better, as we get all the details of this function from its name.

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 1 above)
Comment (danleifeng) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: done.Thanks!

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 2 above)
Comment (danleifeng) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: Done. Thanks!

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 3]
```diff
@@ -115,6 +115,66 @@ def _insert_cast_op(block, op, idx, src_dtype, dest_dtype):
...
+
+    Args:
+        block: The global_block of main program. eg: block = main_prog.global_block()
+    """
+    OPTIMIZE = core.op_proto_and_checker_maker.OpRole.Optimize
+    optimize_ops = []
+    for idx, op in reversed(list(enumerate(block.ops))):
+        if op.attr('op_role') & int(OPTIMIZE):
+            for output_name in op.output_arg_names:
+                if check_next_input_backward_role_op(block.ops, idx, op,
```
Comment (gavin1332) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: why not check all output vars in one call?

Comment (danleifeng) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: Done. Thanks!

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 4]
```diff
@@ -115,6 +115,59 @@ def _insert_cast_op(block, op, idx, src_dtype, dest_dtype):
...
+
+    Args:
+        ops (list): A list of ops.
+        idx (int): index of cur_op in ops.
+        cur_op (Operator): Current operator.
+    """
+    BACKWARD = core.op_proto_and_checker_maker.OpRole.Backward
+    for i in range(idx + 1, len(ops)):
+        op = ops[i]
+        if op.attr('op_role') & int(BACKWARD):
```
Comment (gongweibao) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: !optimize

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 5]
```diff
@@ -115,6 +115,59 @@ def _insert_cast_op(block, op, idx, src_dtype, dest_dtype):
...
+    to speed up in Executor.
+
+    Args:
+        block: The global_block of main program. eg: block = main_prog.global_block()
+    """
+    OPTIMIZE = core.op_proto_and_checker_maker.OpRole.Optimize
+    optimize_ops = []
+    for idx, op in reversed(list(enumerate(block.ops))):
+        if op.attr('op_role') & int(OPTIMIZE):
+            if check_op_validation(block.ops, idx, op):
```
Comment (gongweibao) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:155]: Reversed idx?

**Review by gongweibao — CHANGES_REQUESTED** [review]
Need also unit test.

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 5 above)
Comment (danleifeng) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:155]: 正向remove时会出错

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 4 above)
Comment (danleifeng) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: 只是反向role才判断，这里optimize role不会进来

**Comment by danleifeng** [comment]
> Need also unit test.

test_image_classification_fp16.py调用decorate minize时会覆盖到这部分代码。

**Comment by gongweibao** [comment]
test_image_classification_fp16.py走的逻辑Graph的逻辑，和这个PR的预期场景并不是一致的。建议为这两个改动单独加一个单测。

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 6]
```diff
@@ -115,6 +115,59 @@ def _insert_cast_op(block, op, idx, src_dtype, dest_dtype):
...
+    If no, return True to continue moving in _move_optimize_ops_back.
+
+    Args:
+        ops (list): A list of ops.
+        idx (int): index of cur_op in ops.
+        cur_op (Operator): Current operator.
+    """
+    OPTIMIZE = core.op_proto_and_checker_maker.OpRole.Optimize
+    for i in range(idx + 1, len(ops)):
+        op = ops[i]
```
Comment (gavin1332) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:133]: an alternative way is `for op in ops[idx+1:]`

Comment (danleifeng) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:133]: thanks!

**Pull request closed without merging**